### PR TITLE
List templates in the deployment flow via the API.

### PR DESCRIPTION
**Pull request by kadams54**

### CODE REVIEW

**Comment by kadams54** [comment]
## QA

1. Update your `development.ini` to set the `jujugui.jem_url` setting to `https://api.staging.jujucharms.com/jem/` - the setting is already in there, you'll just need to uncomment it and comment out the line that sets it to nothing.
2. Fire up the GUI with the `blues` flag enabled.
3. Try to deploy something. You should get kicked into the deployment flow.
4. Everyone in the yellow team should have access to, and hence see listed, under "Your cloud credentials", the `kadams54/amazon` template.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4740/
Test PASSed.

File: jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js [hunk 1]
```diff
@@ -33,7 +33,20 @@ YUI.add('deployment-choose-cloud', function() {
...
+
+    getInitialState: function() {
+      return {
+        credentials: []
+      };
+    },
+
+    componentWillMount: function() {
+      this.props.jem.listTemplates((error, data) => {
+        this.setState({credentials: data});
```
Comment (jaycee) [inline jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js]: Shouldn't this do something reasonable in the error state? console.error and return, or something?

File: jujugui/static/gui/src/app/components/deployment/choose-cloud/test-choose-cloud.js [hunk 2]
```diff
@@ -32,20 +32,25 @@ describe('DeploymentChooseCloud', function() {
...
 
   afterEach(function() {
-    juju.components.DeploymentChooseCloud.prototype.CREDENTIALS = credentials;
+    jem = null;
   });
 
   it('can render without credentials', function() {
-    juju.components.DeploymentChooseCloud.prototype.CREDENTIALS = [];
+    jem.listTemplates = (callback) => {
+      callback(null, []);
```
Comment (jaycee) [inline jujugui/static/gui/src/app/components/deployment/choose-cloud/test-choose-cloud.js]: Can you verify that if there are no templates in jem it returns []? I seem to recall that it returned null when we were testing.

**Comment by jaycee** [comment]
Thanks @kadams54, this looks pretty good. I have 2 comments in the diff--in particular please address the error state. We've not done a great job of handling errors when integrating services and it's part of what has led to some fragility.

**Comment by kadams54** [comment]
@jaycee Good points, both. I fixed them both, and added a test specifically for the error case.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4741/
Test PASSed.

**Comment by jaycee** [comment]
@kadams54 thanks for those changes! :+1: 

I'll QA soon.

**Comment by jaycee** [comment]
QA:

I saw my credentials and the ones @kadams54 created available on the page.

QA OK.

File: jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js [hunk 3]
```diff
@@ -22,18 +22,25 @@ YUI.add('deployment-choose-cloud', function() {
...
+    },
+
+    getInitialState: function() {
+      return {
+        credentials: []
+      };
+    },
+
+    componentWillMount: function() {
+      this.props.jem.listTemplates((error, data) => {
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js]: If data is credentials, let's just call it credentials I guess.

File: jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js [hunk 4]
```diff
@@ -22,18 +22,25 @@ YUI.add('deployment-choose-cloud', function() {
...
+        if (error) {
+          console.error('Unable to list templates', error);
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js]: I think this requires at least an XXX comment: we need to handle this error case better I guess.

**Comment by frankban** [comment]
:+1: with minor comments. Thank you!

File: jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js (hunk 4 above)
Comment (kadams54) [inline jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js]: Agreed. I already have a card to polish the UI on this after we finish the "create template" work, which would be a good time to turn this into a user-facing error message. Comment added.

File: jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js (hunk 3 above)
Comment (kadams54) [inline jujugui/static/gui/src/app/components/deployment/choose-cloud/choose-cloud.js]: Yeah, my initial thought was that some data munging was going to be needed to transform data into credentials, but that turned out to not be the case. Variable names updated.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4743/
Test PASSed.

**Comment by kadams54** [comment]
Thanks all! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4744/
Test FAILed.

**Pull request merged**